### PR TITLE
Removed GNU strdup extension from JSON

### PR DESCRIPTION
**Pull request by WillAyd**
Came up during some of the meson work @lithomas1 is doing. Looks like we allow non-portable GNU extensions in JSON, but these aren't upstream in ultrajson. Couldn't find a history of why we have this.

### CODE REVIEW

**Comment by WillAyd** [comment]
See also https://stackoverflow.com/questions/5582211/what-does-define-gnu-source-imply/5583764#5583764

**Review by lithomas1 — APPROVED** [review]
Thanks. I guess we were really out of date on our vendored ujson.

**Comment by WillAyd** [comment]
We are pretty far diverged. Unfortunately when we vendored stuff in the past we would edit in place rather than bolt on to it, so it's hard to sync up with more recent versions

**Comment by WillAyd** [comment]
OK looks like this goes back to https://github.com/pandas-dev/pandas/issues/4918 though I think the strdup calls are unnecessary. Let's see...

File: pandas/_libs/src/ujson/lib/ultrajsonenc.c
Comment (lithomas1) [inline pandas/_libs/src/ujson/lib/ultrajsonenc.c:1180]: Is this error message still correct?

Comment (WillAyd) [reply pandas/_libs/src/ujson/lib/ultrajsonenc.c:1180]: Oh nice catch. I'll change that later

File: pandas/_libs/src/ujson/lib/ultrajsonenc.c
Comment (lithomas1) [inline pandas/_libs/src/ujson/lib/ultrajsonenc.c:1178]: It turns out setlocale can overwrite this string on subsequent calls, which we are doing with ``setlocale(LC_NUMERIC, "C")``.  I guess the strdup is necessary then(although we might be able to replace with malloc/memcpy).

Reference: https://www.gnu.org/software/libc/manual/html_node/Setting-the-Locale.html#:~:text=The%20string%20returned%20by%20setlocale%20can%20be%20overwritten%20by%20subsequent%20calls%2C%20so%20you%20should%20make%20a%20copy%20of%20the%20string%20(see%20Copying%20Strings%20and%20Arrays)%20if%20you%20want%20to%20save%20it%20past%20any%20further%20calls%20to%20setlocale.%20(The%20standard%20library%20is%20guaranteed%20never%20to%20call%20setlocale%20itself.)

**Review by lithomas1 — CHANGES_REQUESTED** [review]

File: pandas/_libs/src/ujson/lib/ultrajsonenc.c
Comment (WillAyd) [reply pandas/_libs/src/ujson/lib/ultrajsonenc.c:1178]: Also a nice find. Cool can copy without strdup

File: pandas/_libs/src/ujson/lib/ultrajsondec.c
Comment (WillAyd) [inline pandas/_libs/src/ujson/lib/ultrajsondec.c:1177]: Not sure when / how this would happen but our current implementation doesn't check for a successful setlocale call. If that ends up being NULL we have undefined behavior in the subsequent strcmp

File: pandas/_libs/src/ujson/lib/ultrajsonenc.c
Comment (lithomas1) [inline pandas/_libs/src/ujson/lib/ultrajsonenc.c:1185]:
```suggestion
        int len = strlen(locale) + 1;
```

**Comment by WillAyd** [comment]
Interestingly enough this becomes part of the standard in C2X

**Comment by lithomas1** [comment]
FYI, the docs build seems to be segfaulting from a quick glance, since it looks like the same thing that happened to me. If you run it with a single worker, you'll be able to see the exit code, I think (245 is segfault). You can also run it with catchsegv, which'll print out more details if it's actually a segfault.

**Comment by WillAyd** [comment]
Super strange. Cool thanks for the heads up - didn't know about that feature

**Comment by WillAyd** [comment]
Looks like I was freeing the wrong variable. Surprised it didn't segfault on other parts of CI - its possible the locale test isn't really even running but will leave that to another effort

cpplint also complains about strcpy but is fine with memcpy...I think the former would be more clear in this case. Worth reconsidering if we want to continue using cpplint given its not well maintained by Google

**Review by lithomas1 — APPROVED** [review]
LGTM. Thanks for cleaning this up. I'll merge this is a day or too, if no one else has any comments.

**Review by mroeschke — APPROVED** [review]

**Comment by mroeschke** [comment]
Nice cleanup @WillAyd

**Pull request merged**
